src/Chapter3.hs

```diff
@@ -1006,6 +1148,77 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+data KnightAction
+  = KAttack
+  | KUseHealPotion
+  | KUseDefenseSpeell
+  deriving (Show)
+
+data MonsterAction
+  = MAttack
+  | MRunAway
+  deriving (Show)
+
+data Knight = Knight
+  { knightName :: String
+  , knightHealth :: Int
+  , knightAttack :: Int
+  , knightDefense :: Int
+  , knightActions :: [KnightAction]
+  } deriving (Show)
+
+data Monster = Monster
+  { monsterName :: String
+  , monsterHealth :: Int
+  , monsterAttack :: Int
+  , monsterActions :: [MonsterAction]
+  } deriving (Show)
+
+class Action a
+instance Action KnightAction
+instance Action MonsterAction
+
+class (Action b) => Fighter a b where
+  getName   :: a -> String
+  getHealth :: a -> Int
+  getNextAction :: a -> (a, b)
+
+instance Fighter Knight KnightAction where
```

Comment:
Yeah, I see what you tried to do. You wanted to strictly connect a creature and its actions on the type level.
This is quite an advanced task, and we didn't cover anything about that to implement it. I am not sure how familiar are you with Haskell, but if would like to implement something like this, you would need to create a custom type family and in the fight function restricts it via the constraint using the type family 🙂 

But generally, in the task we don't require this to be implemented, and you can just pass different actions list into the fight function as an option 🤔

Reply:
Ok, I'll spend some time to learn about type family and try to implement this stuff

Reply:
type family defeated me 🙂